File: 98-validateBST.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param {TreeNode} root
    # @return {boolean}
    # check the upper & lower blund method
    def isValidBST(self, root):
        inf = float('inf')
        return self.bstCheck(root, -inf, inf)
        
    def bstCheck(self, node, past_min, past_max):
        if node is None: return True
        if not past_min < node.val < past_max:
           return False
        return self.bstCheck(node.right, node.val, past_max) and self.bstCheck(node.left, past_min, node.val)

File: C++/98-ValidateBST.cpp
#include "include/common.h"
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool isValidBST(TreeNode* root) {
         double inf = numeric_limits<double>::infinity();
         return validTraverse(root, -inf, inf);
    }
private:
    // Chris:NTR!! must remember!!!
    int startVal;
    bool validTraverse(TreeNode* root, double pastMin, double pastMax) {
         if (root == nullptr) {
             return true;
         };
         if (!(root->val > pastMin && root->val < pastMax)) {
             return false;
         };
         return validTraverse(root->left, pastMin, root->val) && validTraverse(root->right, root->val, pastMax);
    }
};

int main(void) {
    Solution s;
    TreeNode root(0);
    cout << s.isValidBST(&root) << endl;
    return 0;
};